common/scala/src/main/scala/whisk/core/database/cosmosdb/CosmosDBArtifactStoreProvider.scala

```diff
@@ -0,0 +1,104 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.database.cosmosdb
+
+import java.io.Closeable
+
+import akka.actor.ActorSystem
+import akka.stream.ActorMaterializer
+import com.microsoft.azure.cosmosdb.rx.AsyncDocumentClient
+import spray.json.RootJsonFormat
+import whisk.common.Logging
+import whisk.core.database._
+import pureconfig._
+import whisk.core.ConfigKeys
+import whisk.core.database.cosmosdb.CosmosDBUtil.createClient
+import whisk.core.entity.{DocumentReader, WhiskActivation, WhiskAuth, WhiskEntity}
+
+import scala.reflect.ClassTag
+
+case class CosmosDBConfig(endpoint: String, key: String, db: String)
+
+case class ClientHolder(client: AsyncDocumentClient) extends Closeable {
+  override def close(): Unit = client.close()
+}
+
+object CosmosDBArtifactStoreProvider extends ArtifactStoreProvider {
+  type DocumentClientRef = ReferenceCounted[ClientHolder]#CountedReference
+  private lazy val config = loadConfigOrThrow[CosmosDBConfig](ConfigKeys.cosmosdb)
+  private var clientRef: ReferenceCounted[ClientHolder] = _
+
+  override def makeStore[D <: DocumentSerializer: ClassTag](useBatching: Boolean)(
+    implicit jsonFormat: RootJsonFormat[D],
+    docReader: DocumentReader,
+    actorSystem: ActorSystem,
+    logging: Logging,
+    materializer: ActorMaterializer): ArtifactStore[D] = {
+    makeStoreForClient(config, useBatching, getOrCreateReference(config))
+  }
+
+  def makeStore[D <: DocumentSerializer: ClassTag](config: CosmosDBConfig, useBatching: Boolean)(
+    implicit jsonFormat: RootJsonFormat[D],
+    docReader: DocumentReader,
+    actorSystem: ActorSystem,
+    logging: Logging,
+    materializer: ActorMaterializer): ArtifactStore[D] = {
+
+    makeStoreForClient(config, useBatching, createReference(config).reference())
+  }
+
+  private def makeStoreForClient[D <: DocumentSerializer: ClassTag](config: CosmosDBConfig,
+                                                                    useBatching: Boolean,
+                                                                    clientRef: DocumentClientRef)(
+    implicit jsonFormat: RootJsonFormat[D],
+    docReader: DocumentReader,
+    actorSystem: ActorSystem,
+    logging: Logging,
+    materializer: ActorMaterializer): ArtifactStore[D] = {
+
+    val classTag = implicitly[ClassTag[D]]
+    val (dbName, handler, viewMapper) = handlerAndMapper(classTag)
+
+    new CosmosDBArtifactStore(dbName, config, clientRef, handler, viewMapper)
+  }
+
+  private def handlerAndMapper[D](entityType: ClassTag[D])(
+    implicit actorSystem: ActorSystem,
+    logging: Logging,
+    materializer: ActorMaterializer): (String, DocumentHandler, CosmosDBViewMapper) = {
+    entityType.runtimeClass match {
+      case x if x == classOf[WhiskEntity] =>
```

Comment:
is there a better way? does ` case x:WhiskEntity =>` not work?

Reply:
That would work if x is an instance of `WhiskEntity`. However here x is class representing `WhiskEntity`. So current approach has to be taken

Reply:
given the matching and `if` predicates, might as well use if/else directly for brevity.